### PR TITLE
[DRIVER] bmp180 sensor driver implementation

### PR DESCRIPTION
**Pull request by aabadie**
Hi,
I have a [bmp180](https://www.adafruit.com/products/1603) and noticed that there's no driver implementation for it.
I tested this implementation using a Samr21-xpro (SDA is pin PA16, SCL is PA17). It works well !

### CODE REVIEW

File: drivers/bmp180/bmp180.c [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    printf("AC3: %i\n", (int)dev->calibration.ac3);
+    printf("AC4: %i\n", (int)dev->calibration.ac4);
+    printf("AC5: %i\n", (int)dev->calibration.ac5);
+    printf("AC6: %i\n", (int)dev->calibration.ac6);
+    printf("B1: %i\n",  (int)dev->calibration.b1);
+    printf("B2: %i\n",  (int)dev->calibration.b2);
+    printf("MB: %i\n",  (int)dev->calibration.mb);
+    printf("MC: %i\n",  (int)dev->calibration.mc);
+    printf("MD: %i\n",  (int)dev->calibration.md);
+#endif
```
Comment (kYc0o) [inline drivers/bmp180/bmp180.c]: Can you use the `DEBUG()` function available on `debug.h`?

Comment (aabadie) [inline drivers/bmp180/bmp180.c]: @kYc0o : done !

**Comment by aabadie** [comment]
Just a side note about potential improvements: section 3.6 and 3.7 of the [bmp180 datasheet](https://www.adafruit.com/datasheets/BST-BMP180-DS000-09.pdf) explain how to compute respectively the absolute altitude and pressure at see level. 

Maybe I can add the 2 missing functions to compute those values ?

**Comment by kYc0o** [comment]
Is up to you! the driver can offer basic functionality but if you want to implement all the features it's always better

File: drivers/bmp180/bmp180.c [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    DEBUG("AC3: %i\n", (int)dev->calibration.ac3);
+    DEBUG("AC4: %i\n", (int)dev->calibration.ac4);
+    DEBUG("AC5: %i\n", (int)dev->calibration.ac5);
+    DEBUG("AC6: %i\n", (int)dev->calibration.ac6);
+    DEBUG("B1: %i\n",  (int)dev->calibration.b1);
+    DEBUG("B2: %i\n",  (int)dev->calibration.b2);
+    DEBUG("MB: %i\n",  (int)dev->calibration.mb);
+    DEBUG("MC: %i\n",  (int)dev->calibration.mc);
+    DEBUG("MD: %i\n",  (int)dev->calibration.md);
+#endif
```
Comment (kYc0o) [inline drivers/bmp180/bmp180.c]: then the `#if ENABLE_DEBUG` becomes unnecessary don't you think?

Comment (aabadie) [inline drivers/bmp180/bmp180.c]: Indeed, I removed them.

**Comment by aabadie** [comment]
> Is up to you! the driver can offer basic functionality but if you want to implement all the features it's always better

I added the missing functions.
Just one side question: is it possible to display `float` values using `printf`function ? I tried `%f` but nothing is displayed. Maybe you have an idea.

**Comment by aabadie** [comment]
> Maybe you have an idea.

Answering to myself: apparently using float_t type is the way to go.

**Comment by gebart** [comment]
`%f` requires support in newlib, you're likely building with newlib-nano. Float printf support can be added with some linker flags though I can't remember which right now.

**Comment by aabadie** [comment]
> %f requires support in newlib, you're likely building with newlib-nano

Indeed, but I don't know how to enable this either.

**Comment by aabadie** [comment]
> Indeed, but I don't know how to enable this either.

It seems that I have to apply RIOT [patches](https://github.com/RIOT-OS/RIOT/tree/master/dist/tools/toolchains/patches) on newlib sources, rebuild them and re-install them, using the available scripts.
Is there any documentation somewhere ?

**Comment by gebart** [comment]
@aabadie I don't think you need to rebuild newlib. I did this in Contiki before to get float printf support:
https://github.com/eistec/contiki/commit/b5d6b25b4415c3100a2335b7dacf1974524bc03c

**Comment by aabadie** [comment]
Thanks @gebart, I tried to add `LDFLAGS += -specs=nano.specs -lc -lnosys -u _printf_float` in the `Makefile` for the test application but still no float values displayed.

**Comment by gebart** [comment]
@aabadie Try adding it to the end of LINKFLAGS at https://github.com/RIOT-OS/RIOT/blob/master/cpu/Makefile.include.cortexm_common#L124

**Comment by aabadie** [comment]
Thanks @gebart, this time it works !
I pushed an updated version of the test application.

**Comment by kYc0o** [comment]
Maybe we need to take care about the platforms that can support it, since not all can handle floats in this way. Let's run the CI and if everything goes ok we can merge it ;)

**Comment by kYc0o** [comment]
Oh! it as already complaining about the white spaces. Can you take care of this?

File: cpu/Makefile.include.cortexm_common [hunk 3]
```diff
@@ -95,7 +95,7 @@ include $(RIOTCPU)/cortexm_common/Makefile.include
 
 # use the nano-specs of Newlib when available
 ifeq ($(shell $(LINK) -specs=nano.specs -E - 2>/dev/null >/dev/null </dev/null ; echo $$?),0)
-export LINKFLAGS += -specs=nano.specs -lc -lnosys
+export LINKFLAGS += -specs=nano.specs -lc -lnosys -u _printf_float
```
Comment (LudwigKnuepfer) [inline cpu/Makefile.include.cortexm_common]: I would suggest to split this out into a separate pull request.

**Comment by LudwigKnuepfer** [comment]
`tests/driver_bmp180/Readme.md` is empty

File: tests/driver_bmp180/main.c [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+#endif
+
+#include <stdio.h>
+#include <inttypes.h>
+
+#include "bmp180.h"
+#include "xtimer.h"
+#include "board.h"
+
+#define SLEEP   (2 * 1000 * 1000u)
```
Comment (LudwigKnuepfer) [inline tests/driver_bmp180/main.c]: maybe rename to `SLEEP_2MS` or document..

Comment (LudwigKnuepfer) [inline tests/driver_bmp180/main.c]: (2S or whatever this is ;)

File: cpu/Makefile.include.cortexm_common (hunk 3 above)
Comment (aabadie) [inline cpu/Makefile.include.cortexm_common]: Added #5094 for this.

File: tests/driver_bmp180/Makefile [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+FEATURES_REQUIRED = periph_i2c
+
+USEMODULE += bmp180
+USEMODULE += xtimer
+
+# define parameters for selected boards
+ifneq (,$(filter msbiot,$(BOARD)))
+  TEST_I2C       ?= I2C_0
+  TEST_MEASURE_OVERSAMPLING ?= BMP180_ULTRALOWPOWER
+endif
```
Comment (LudwigKnuepfer) [inline tests/driver_bmp180/Makefile]: seems redundant (same as below)

File: tests/driver_bmp180/main.c [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        /* Get temperature in degrees celsius */
+        bmp180_read_temperature(&dev, &temperature);
+        printf("Temperature [°C] : %.1f\n", (double)temperature / 10.0);
+
+        /* Get pressure in hPa */
+        bmp180_read_pressure(&dev, &pressure);
+        printf("Pressure [hPa] : %.2f\n", (double)pressure / 100.0);
+
+        /* Get pressure at sealevel in hPa */
+        double x_altitude = 158; // Altitude in Polytechnique campus
```
Comment (LudwigKnuepfer) [inline tests/driver_bmp180/main.c]: Maybe parametrize this via Makefile?

File: drivers/include/bmp180.h [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+ * @name BMP180 registers
+ * @{
+ */
+#define BMP180_REGISTER_ID            (0xD0)
+#define BMP180_REGISTER_CONTROL       (0xF4)
+#define BMP180_REGISTER_DATA          (0xF6)
+#define BMP180_TEMPERATURE_COMMAND    (0x2E)
+#define BMP180_PRESSURE_COMMAND       (0x34)
+#define BMP180_CALIBRATION_AC1        (0xAA)
+/** @} */
```
Comment (LudwigKnuepfer) [inline drivers/include/bmp180.h]: These defines should not be part of the public header.

Comment (LudwigKnuepfer) [inline drivers/include/bmp180.h]: Look at some other drivers which use a pattern like `drivers/include/foo.h` and `drivers/foo/include/foo-internal.h`

File: tests/driver_bmp180/main.c [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        bmp180_sealevel_pressure(&dev, &x_altitude, &pressure_0);
+        printf("Pressure at see level [hPa] : %.2f\n", pressure_0 / 100.0);
+
+        /* Get altitude in meters */
+        bmp180_altitude(&dev, &pressure_0, &altitude);
+        printf("Altitude [m] : %.1f\n", altitude);
```
Comment (LudwigKnuepfer) [inline tests/driver_bmp180/main.c]: Feeding the above altitude back into the device is interesting but not really sensible .. all it will do is fluctuate around the altitude level given in `x_altitude`.

**Comment by LudwigKnuepfer** [comment]
Tested, works.

File: tests/driver_bmp180/main.c (hunk 8 above)
Comment (aabadie) [inline tests/driver_bmp180/main.c]: The idea behind is that given the current altitude, the driver provides a helper function computing the pressure at sea level and vice-versa. This corresponds to sections 3.6 and 3.7 of the [datasheet](https://www.adafruit.com/datasheets/BST-BMP180-DS000-09.pdf).

**Comment by aabadie** [comment]
@LudwigKnuepfer, I addressed most of your comments.

File: tests/driver_bmp180/main.c (hunk 8 above)
Comment (LudwigKnuepfer) [inline tests/driver_bmp180/main.c]: Yes, I was just wondering if it's not more to the point to provide a reference pressure to calculate the altitude.. Otherwise it's just adding uncertainty twice - once for the "forward" and once for the "backward" operation ;)

File: drivers/include/bmp180.h [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+ * @brief Compute altitude, returned in m.
+ *
+ * @param[in]  dev          Device descriptor of BMP180 device to read from
+ * @param[in]  pressure_0   The pressure at sea level in Pa
+ * @param[out] altitude     Altitude in m
+ *
+ * @return                  0 on success
+ * @return                  -1 if device's I2C is not enabled in board config
+ */
+int bmp180_altitude(bmp180_t *dev, double *pressure_0, double *altitude);
```
Comment (haukepetersen) [inline drivers/include/bmp180.h]: I am not sure if double is a good idea here - I would strongly vote for something more ressource friendly - why not altitute in cm as `int32_t`? I doubt you get a better precision anyway... Same for the pressure...

File: drivers/bmp180/include/bmp180_internals.h [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name BMP180 I2C address
+ * @{
+ */
+#define BMP180_ADDR                   (0x77) // 7 bits address
```
Comment (haukepetersen) [inline drivers/bmp180/include/bmp180_internals.h]: use c-style comment. And could you add the alignment of the 7-bit address (right as it seems)?! thx

File: drivers/include/bmp180.h (hunk 9 above)
Comment (aabadie) [inline drivers/include/bmp180.h]: Ok, altitude precision can be 1m, under I think it doesn't really make sense. I changed the returned type from `double` to `int32_t`.

File: drivers/include/bmp180.h [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+ * @brief Compute altitude, returned in m.
+ *
+ * @param[in]  dev          Device descriptor of BMP180 device to read from
+ * @param[in]  pressure_0   The pressure at sea level in Pa
+ * @param[out] altitude     Altitude in m
+ *
+ * @return                  0 on success
+ * @return                  -1 if device's I2C is not enabled in board config
+ */
+int bmp180_altitude(bmp180_t *dev, int32_t *pressure_0, int32_t *altitude);
```
Comment (haukepetersen) [inline drivers/include/bmp180.h]: pressure_0 is an input, right? So why pass it as a pointer?

Comment (aabadie) [inline drivers/include/bmp180.h]: No particular reason (seems to be a bad practice ;) ). I changed that.

**Comment by kYc0o** [comment]
Murdock is complaining about `nhdp` tests... otherwise all other tests passed successfully. @kaspar030 is this related to the move of `uhdp` out of the examples?

**Comment by gebart** [comment]
@kYc0o no, it's because of https://github.com/RIOT-OS/RIOT/issues/5107

File: tests/driver_bmp180/Readme.md [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+coefficients. Please see section 3.4 of the
+[datasheet](https://www.adafruit.com/datasheets/BST-BMP180-DS000-09.pdf) for
+more information.
+
+After initialization, every 2 seconds, the application:
+* reads temperature (°C);
+* reads pressure (hPa) ; and prints them to the STDOUT;
+* computes the pressure at sea level (assuming the current pressure is computed
+based on Polytechnique School Campus altitude, 158m);
+* computes the altitude (given the pressure at sea level).
```
Comment (LudwigKnuepfer) [inline tests/driver_bmp180/Readme.md]: `given the pressure at sea level` => `based on the pressure at sea level calculated above`

Comment (aabadie) [inline tests/driver_bmp180/Readme.md]: > `given the pressure at sea level` => `based on the pressure at sea level
> calculated above`

branch updated including this change.

**Comment by aabadie** [comment]
Just pushed a commit activating `printf_float` module in test application (see #5094).

File: drivers/include/bmp180.h [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+ * @return                  0 on success
+ * @return                  -1 if device's I2C is not enabled in board config
+ */
+int bmp180_read_temperature(bmp180_t *dev, int32_t *temperature);
+
+/**
+ * @brief Read pressure value from the given BMP180 device, returned in Pa
+ *
+ * @param[in]  dev          Device descriptor of BMP180 device to read from
+ * @param[out] pressure     Pressure in Pa
```
Comment (haukepetersen) [inline drivers/include/bmp180.h]: I think changing the output range to use 10^-1000 Pa (mPa) is probably more precise and feasible. If I am correct, the pressure is somewhere between 10 and 150 from sea-level to very high up (https://upload.wikimedia.org/wikipedia/commons/thumb/8/88/Atmospheric_Pressure_vs._Altitude.png/300px-Atmospheric_Pressure_vs._Altitude.png), so using mPa here would be more precise...

Comment (aabadie) [inline drivers/include/bmp180.h]: > I think changing the output range to use 10^-1000 Pa (mPa) is probably more
> precise and feasible. If I am correct, the pressure is somewhere between 10
> and 150 from sea-level to very high up
> (https://upload.wikimedia.org/wikipedia/commons/thumb/8/88/Atmospheric_Pressure_vs._Altitude.png/300px-Atmospheric_Pressure_vs._Altitude.png),
> so using mPa here would be more precise...

I'm not sure to understand your comment, maybe you mean kPa ? the results here are already returned in Pa, which is 1000x more precise and corresponds to SI. Displaying the results in hPa is common in meteorology.

**Comment by kYc0o** [comment]
All the checks are passed, someone with the hardware can ACK? @LudwigKnuepfer ?

File: drivers/include/bmp180.h (hunk 13 above)
Comment (haukepetersen) [inline drivers/include/bmp180.h]: sorry, my bad, I was way of here... Don't know what I was thinking - never mind...

Comment (aabadie) [inline drivers/include/bmp180.h]: np. 
Thanks to your comment I had to look again at the datasheet and this pointed me to the possibility of writing a unittest for the temperature/pressure computations: the datasheet gives ground truth values that could be used to test internal implementations of formulas in the driver.
Do you think it make sense ?

**Comment by LudwigKnuepfer** [comment]
it would be great if you could rewrite your commit messages to fit the `module: change` pattern (and still fit within the 70 character limitation for the first line of the commit message).

**Comment by LudwigKnuepfer** [comment]
Example: `drivers/bmp180: initial import`

**Comment by aabadie** [comment]
@LudwigKnuepfer, I edited the commit message following your suggestion.

**Comment by LudwigKnuepfer** [comment]
ACK

**Comment by aabadie** [comment]
> Merged #5078.

Thanks :)Merged #5078.

---
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub:
https://github.com/RIOT-OS/RIOT/pull/5078#event-599203907

**Comment by LudwigKnuepfer** [comment]
No, thank *you* for your contribution and endurance with all the review comments =)
